Libraries/LibGfx/JPGLoader.cpp

```diff
@@ -0,0 +1,1057 @@
+    dbg() << "Macroblocks in a column: " << context.mblock_meta.vpadded_count;
+
+    // Compute huffman codes for DC and AC tables.
+    for (auto& dc_table : context.dc_tables)
+        generate_huffman_codes(dc_table);
+
+    for (auto& ac_table : context.ac_tables)
+        generate_huffman_codes(ac_table);
+
+    for (u32 vcursor = 0; vcursor < context.mblock_meta.vcount; vcursor += context.vsample_factor) {
+        for (u32 hcursor = 0; hcursor < context.mblock_meta.hcount; hcursor += context.hsample_factor) {
+            u32 i = vcursor * context.mblock_meta.hpadded_count + hcursor;
+            if (context.dc_reset_interval > 0) {
+                if (i % context.dc_reset_interval == 0) {
+                    context.previous_dc_values[0] = 0;
+                    context.previous_dc_values[1] = 0;
+                    context.previous_dc_values[2] = 0;
+
+                    // Restart markers are stored in byte boundaries. Advance the huffman stream cursor to
+                    //  the 0th bit of the next byte.
+                    if (context.huffman_stream.byte_offset < context.huffman_stream.stream.size()) {
+                        if (context.huffman_stream.bit_offset > 0) {
+                            context.huffman_stream.bit_offset = 0;
+                            context.huffman_stream.byte_offset++;
+                        }
+
+                        // Skip the restart marker (RSTn).
+                        context.huffman_stream.byte_offset++;
+                    }
+                }
+            }
+
+            if (!build_macroblocks(context, macroblocks, hcursor, vcursor)) {
+                dbg() << "Failed to build Macroblock " << i + 1;
+                dbg() << "Huffman stream byte offset " << context.huffman_stream.byte_offset;
+                dbg() << "Huffman stream bit offset " << context.huffman_stream.bit_offset;
+                return {};
+            }
+        }
+    }
+
+    return macroblocks;
+}
+
+static inline bool bounds_okay(const size_t cursor, const size_t delta, const size_t bound)
+{
+    return (delta + cursor) < bound;
+}
+
+static inline bool is_valid_marker(const Marker marker)
+{
+    if (marker >= JPG_APPN0 && marker <= JPG_APPNF) {
+        if (marker != JPG_APPN0)
+            dbg() << String::format("%04x not supported yet. The decoder may fail!", marker);
+        return true;
+    }
+    if (marker >= JPG_RESERVED1 && marker <= JPG_RESERVEDD)
+        return true;
+    if (marker >= JPG_RST0 && marker <= JPG_RST7)
+        return true;
+    switch (marker) {
+    case JPG_COM:
+    case JPG_DHP:
+    case JPG_EXP:
+    case JPG_DHT:
+    case JPG_DQT:
+    case JPG_RST:
+    case JPG_SOF0:
+    case JPG_SOI:
+    case JPG_SOS:
+        return true;
+    default:
+        return false;
+    }
+}
+
+static inline u16 read_endian_swapped_word(BufferStream& stream)
+{
+    u16 temp { 0 };
+    stream >> temp;
+    return (temp >> 8) | (temp << 8);
+}
```

Comment:
Since everything in the JPEG is big-endian (I think, at least according to T.81 Appendix B. section 1.1.1), renaming the method to `read_be_word` and enforcing reading a BE word via `stream >> temp0; stream >> temp1; return (((u16)temp0) << 8) | ((u16)temp1);` would be more appropriate.
I don't know if Serenity is supposed to be able to run on big-endian machines, but we might as well make sure that if it does then JPG still works ^^